desc/optimize/tr_subproblems.py

```diff
@@ -370,6 +376,118 @@ def loop_body(state):
     return cond(jnp.linalg.norm(p_newton) <= trust_radius, truefun, falsefun, None)
 
 
+@jit
+def trust_region_step_exact_qr(
+    f, J, trust_radius, initial_alpha=None, rtol=0.01, max_iter=10
+):
+    """Solve a trust-region problem using a semi-exact method.
+
+    Solves problems of the form
+        min_p ||J*p + f||^2,  ||p|| < trust_radius
+
+    Parameters
+    ----------
+    f : ndarray
+        Vector of residuals.
+    J : ndarray
+        Jacobian matrix.
+    trust_radius : float
+        Radius of a trust region.
+    initial_alpha : float, optional
+        Initial guess for alpha, which might be available from a previous
+        iteration. If None, determined automatically.
+    rtol : float, optional
+        Stopping tolerance for the root-finding procedure. Namely, the
+        solution ``p`` will satisfy
+        ``abs(norm(p) - trust_radius) < rtol * trust_radius``.
+    max_iter : int, optional
+        Maximum allowed number of iterations for the root-finding procedure.
+
+    Returns
+    -------
+    p : ndarray, shape (n,)
+        Found solution of a trust-region problem.
+    hits_boundary : bool
+        True if the proposed step is on the boundary of the trust region.
+    alpha : float
+        Positive value such that (J.T*J + alpha*I)*p = -J.T*f.
+        Sometimes called Levenberg-Marquardt parameter.
+
+    """
+    # try full newton step
+    tall = J.shape[0] >= J.shape[1]
+    if tall:
+        Q, R = qr(J, mode="economic")
+        p_newton = solve_triangular_regularized(R, -Q.T @ f)
+    else:
+        Q, R = qr(J.T, mode="economic")
+        p_newton = Q @ solve_triangular_regularized(R.T, f, lower=True)
+
+    def truefun(*_):
+        return p_newton, False, 0.0
+
+    def falsefun(*_):
```

Comment:
so this is what's used if there is no exact solution within the trust region? could you give me a general overview of the function? I can't really tell what's going on

Reply:
Basically we first check to see if the full newton step is within the trust region. If it is, we use that (`truefun`).

Otherwise, we do some rootfinding in the regularization parameter `alpha` to find the regularized step that lies on the boundary of the trust region (`falsefun`).